File: 453-flatten-binary-tree-to-linked-list/divide-conquer.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    void flatten(TreeNode * root) {
        // write your code here
        if (root == NULL) {
            return;
        }
        preorderSearch(root);
    }
    TreeNode* preorderSearch(TreeNode* root) {
        if (root == NULL) {
            return NULL;
        }
        TreeNode* leftLast = preorderSearch(root->left);
        TreeNode* rightLast = preorderSearch(root->right);
        if (leftLast != NULL) {
            leftLast->right = root->right;
            root->right = root->left;
            root->left = NULL;
        }
        if (rightLast != NULL) {
            return rightLast;
        }
        if (leftLast != NULL) {
            return leftLast;
        }
        return root;
    }
};

File: 628-maximum-subtree/divide_conquer.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of binary tree
     * @return: the maximum weight node
     */
    private int max = Integer.MIN_VALUE;
    TreeNode result = new TreeNode(0);
    public TreeNode findSubtree(TreeNode root) {
        // write your code here
        if (root == null) {
            return null;
        }
        findHelper(root);
        return result;
    }
    int findHelper(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = findHelper(root.left);
        int right = findHelper(root.right);
        int temp = root.val + left + right;
        if (temp > max) {
            max = temp;
            result = root;
        }
        return temp;
    }
}
